google/cloud/spanner/client_test.cc

```diff
@@ -124,6 +126,93 @@ TEST(SpannerClientTest, ReadTest) {
+            EXPECT_EQ(database_name, request.database());
+            return Status(StatusCode::kPermissionDenied, "uh-oh in GetSession");
+          }));
+  auto commit = client.Commit(MakeReadWriteTransaction(), {});
+  EXPECT_EQ(StatusCode::kPermissionDenied, commit.status().code());
+  EXPECT_THAT(commit.status().message(), HasSubstr("uh-oh in GetSession"));
+}
+
+/// @test Verify that errors when creating sessions are reported.
+TEST(SpannerClientTest, Commit_CommitFailure) {
+  auto mock = std::make_shared<MockSpannerStub>();
+
+  auto database_name =
+      MakeDatabaseName("dummy_project", "dummy_instance", "dummy_database_id");
+  Client client(database_name, mock);
+  EXPECT_CALL(*mock, CreateSession(_, _))
+      .WillOnce(::testing::Invoke(
+          [&database_name](
+              grpc::ClientContext&,
+              google::spanner::v1::CreateSessionRequest const& request) {
+            EXPECT_EQ(database_name, request.database());
+            google::spanner::v1::Session session;
+            session.set_name("test-session-name");
+            return session;
+          }));
+  EXPECT_CALL(*mock, Commit(_, _))
+      .WillOnce(::testing::Invoke(
+          [](grpc::ClientContext&,
+             google::spanner::v1::CommitRequest const& request) {
+            EXPECT_EQ("test-session-name", request.session());
+            return Status(StatusCode::kPermissionDenied, "uh-oh in Commit");
+          }));
+  auto commit = client.Commit(MakeReadWriteTransaction(), {});
+  EXPECT_EQ(StatusCode::kPermissionDenied, commit.status().code());
+  EXPECT_THAT(commit.status().message(), HasSubstr("uh-oh in Commit"));
+}
+
+/// @test Verify that transaction ids are used correctly.
+TEST(SpannerClientTest, Commit_TransactionId) {
+  auto mock = std::make_shared<MockSpannerStub>();
+
+  auto database_name =
+      MakeDatabaseName("dummy_project", "dummy_instance", "dummy_database_id");
+  Client client(database_name, mock);
+  EXPECT_CALL(*mock, CreateSession(_, _))
+      .WillOnce(::testing::Invoke(
+          [&database_name](
+              grpc::ClientContext&,
+              google::spanner::v1::CreateSessionRequest const& request) {
+            EXPECT_EQ(database_name, request.database());
+            google::spanner::v1::Session session;
+            session.set_name("test-session-name");
+            return session;
+          }));
+  EXPECT_CALL(*mock, Commit(_, _))
+      .WillOnce(::testing::Invoke(
+          [](grpc::ClientContext&,
+             google::spanner::v1::CommitRequest const& request) {
+            EXPECT_EQ("test-session-name", request.session());
+            EXPECT_EQ("test-txn-id", request.transaction_id());
+            return Status(StatusCode::kPermissionDenied, "uh-oh in Commit");
+          }));
+
+  auto txn = MakeReadWriteTransaction();
+  using F = std::function<int(spanner_proto::TransactionSelector&)>;
+  internal::Visit(txn, F([](spanner_proto::TransactionSelector& s) -> int {
```

Comment:
Aside: This is like `client.Begin(txn)`.

Reply:
Ack.